Feature: Cookie value serialize

jossemarGT opened the pull request:
I was tempted to add everything on a single branch, but for now this feature only adds the Tornado secure cookie value serialization and de-serialization logic to the library.

By the way, the library only holds types so it is clear is not ready for its first release yet, but I wanted to have feedback.

jossemarGT committed:
Add library initial scope definition

I am writing down the library usage specification and scope as
an user manual, so I could use it as a guide in the solution
design and implementation.

jossemarGT committed:
Add CookieValueModel + Tornado serializer/de-serializer interfaces

Since the Tornado secure cookie values holds so many metadata on
it, it was more simple to represent them as a plain Java object
or a "model", in that way the serialize and de-serialize methods
could be simplified by using it.

jossemarGT committed:
Update internal documentation and update package layout

Since the README file is not enough I updated the javadoc comments
to reflect the ideas behind the implementation and re-organized
the package layout.

jossemarGT committed:
Refactor naming to all the tornado package members

I started this library with common practices from other languages
which means the whole library was not following some common Java
naming conventions and package layout.

PS: All the changes where done in several commits, but I prefered
to squash them for git history readability sake.

jossemarGT commented on build.gradle:
```diff
@@ -23,7 +27,25 @@ repositories {
+
+test {
+    // In the meantime I implement the Library API
+    exclude '**/CookiePotTest*'
```
Since CookiePot is an empty facade its test will always fail.

diegofigueroa replied:
I think you can just annotate the test class with `@Ignore` and achieve the same result :)

jossemarGT replied:
You're right, I just changed it. Thanks

lfmexi commented on src/main/java/com/jossemargt/cookietwist/exception/InvalidFormatException.java:
```diff
@@ -0,0 +1,73 @@
+ * or un-expected amount of String tokens.
+ */
+@SuppressWarnings("serial")
+public class InvalidFormatException extends Exception {
```
Could you please extend your exception from `RuntimeException`? Probably your library's clients would not like to deal with checked exceptions.

jossemarGT replied:
Being honest, I was completely unaware of the difference between checked and unchecked Exceptions until you suggested this, and sure thing!  I'm going to change it and update the class/interface names into a more (not so fancy) Java way as well, based on the off PR conversation that we had during this week.

jossemarGT replied:
Done.

BTW I'm trying to figure out if I should rename the `value` package to `transform` instead, since it contains the serialization objects, but still that new name doesn't sound alright for me. Do you have any suggestion?

jossemarGT commented on build.gradle:
```diff
@@ -23,7 +27,27 @@ repositories {
+    exclude '**/CookiePotTest*'
+}
+
+compileTestJava.options.encoding = 'UTF-8'
```
Because some test cases make use of UTF-8 characters and the wrong enconding will lead to false negatives.

diegofigueroa commented on README.md:
```diff
@@ -3,9 +3,73 @@
+import javax.servlet.http.Cookie;
+import com.jossemargt.cookietwist.CookiePot;
+
+public static void main(string args[]){
```
`string` should be `String`

diegofigueroa replied:
the same happens in other places of this example...

jossemarGT replied:
Good catch, that's a Golang habit, I just changed it. Thanks

diegofigueroa commented on README.md:
```diff
@@ -3,9 +3,73 @@
+
+    // Generate a Tornado Signed Secure Cookie (with signing V2 by default);
+    Cookie secureCookie = tornadoPot.encodeCookie(new Cookie('the_name', 'a value'));
+    string hashedValue = secureCookie.getValue();
```
so the idea is to grab `hashedValue` and send that in the set-cookie header?

jossemarGT replied:
Well, what to do with the cookie is up to the developer, do you think I should add how to set the cookie header as well?

diegofigueroa commented on src/main/java/com/jossemargt/cookietwist/tornado/TornadoCookieValue.java:
```diff
@@ -0,0 +1,268 @@
+     *
+     * @param builder the TornadoCookieValueBuilder
+     */
+    public TornadoCookieValue(TornadoCookieValueBuilder builder) {
```
given you have a builder, should this constructor be made private?

jossemarGT replied:
Indeed, good catch.

diegofigueroa commented on src/main/java/com/jossemargt/cookietwist/tornado/TornadoCookieValue.java:
```diff
@@ -0,0 +1,268 @@
+ * The Class TornadoCookieValue is plain Java representation of a Tornado cookie
+ * value string.
+ */
+public class TornadoCookieValue {
```
have you thought about making this immutable?

jossemarGT replied:
Immutable as making all the Class members final and get rid of the setters?

jossemarGT replied:
After a quick research (I just googled it)  I found  [this short article](https://docs.oracle.com/javase/tutorial/essential/concurrency/immutable.html) explaining more about immutability, which  reminds me of how Golang solves concurrency using a similar principle (everything is a value) as well, so I'm going to follow your suggestion using [this article](https://docs.oracle.com/javase/tutorial/essential/concurrency/imstrat.html).

BTW Thanks, I wouldn't spotted it by myself :+1:

diegofigueroa commented on src/main/java/com/jossemargt/cookietwist/tornado/value/impl/V1TornadoCookieValueSerializer.java:
```diff
@@ -0,0 +1,57 @@
+        StringBuilder result =  new StringBuilder(rawResult);
+
+        if (model.getSignature() != null && !model.getSignature().isEmpty()) {
+            result.append("|").append(model.getSignature());
```
I wouldn't use a `StringBuilder`, as it just makes the code a bit less clear, compared to doing a simple `String` concatenation like `rawResult + "|" + signature`. Also, the compiler will generate the code using the `StringBuilder` for you, based on the simple concatenation.

jossemarGT replied:
I didn't know that the compiler would do such optimizations. Thanks

jossemarGT committed:
Transform TornadoCookieValue into an immutable type

Making the needed changes in order to have an immutable
TornadoCookieValue which makes easier the concurrency handling
in the library client's side.

yhonatan25 commented on src/main/java/com/jossemargt/cookietwist/CookiePot.java:
```diff
@@ -0,0 +1,32 @@
+/*
```
Just curious about why is not enough adding the MIT license on the repo. I mean, why do you need to add it on all your classes?

jossemarGT replied:
I have to honest with you, I just followed the patterns that I have observed in other open source Java projects, like Jenkins and several Jenkins plugins. By the way I did a quick google search about the topic and the tl;dr answer is "to remind about the license and that any extract of the source code is tied to it". 

So I'll stick to it, since those lines can be ignored and does not alter the Java artifact nor the javadoc :man_shrugging: 

PS: Thanks for asking, that pushed me to ask myself about it

jossemarGT commented:
Ok folks, thanks a lot for you comments it helped me to catch up with Java conventions and to improve this small feature. I'm going to merge this PR, but if you spot something else please don't hesitate to leave your comments.
